subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy

```diff
@@ -62,12 +62,60 @@ class JavaReflectionUtilTest extends Specification {
 
     def "write property"() {
         when:
-        writeableProperty(JavaTestSubject, "myProperty").setValue(myProperties, "otherValue")
+        writeableProperty(JavaTestSubject, "myProperty", String.class).setValue(myProperties, "otherValue")
 
         then:
         readableProperty(JavaTestSubject, String, "myProperty").getValue(myProperties) == "otherValue"
     }
 
+    def "write property with multiple setters"() {
+        when:
+        writeableProperty(JavaTestSubject, "myProperty2", String.class).setValue(myProperties, "stringValue")
+
+        then:
+        readableProperty(JavaTestSubject, String, "myProperty2").getValue(myProperties) == "stringValue"
+
+        when:
+        writeableProperty(JavaTestSubject, "myProperty2", File.class).setValue(myProperties, new File("fileValue"))
+
+        then:
+        readableProperty(JavaTestSubject, String, "myProperty2").getValue(myProperties) == "fileValue"
+    }
+
+    def "picks the generic object setter if the typed setter does not match the value type"() {
+        when:
+        def property = writeableProperty(JavaTestSubject, "myProperty", File.class)
+
+        then:
+        property.type == Object.class
+    }
+
+    def "picks the generic iterable setter if the typed setter does not match the value type"() {
+        when:
+        def property = writeableProperty(JavaTestSubject, "multiValue", List.class)
+
+        then:
+        property.type == Iterable.class
+    }
+
+    def "can handle null as property type"() {
+        when:
+        writeableProperty(JavaTestSubject, "myProperty", null)
+
+        then:
+        //we do not know which 'myProperty' setter is picked, as both fit equally well
+        noExceptionThrown()
+    }
+
+    def "cannot write primitive type properties if type is unknown"() {
+        when:
+        writeableProperty(JavaTestSubject, "myBooleanProperty", null)
+
+        then:
+        def e = thrown(NoSuchPropertyException)
+        e.message == "Could not find setter method for property 'myBooleanProperty' on class JavaTestSubject."
```

Comment:
It might be confusing to read that message when you're sure there's a setter method just not one that accepts `null`.

How about mentioning the value was null? Something like:

`Could not find setter method for property 'myBooleanProperty' accepting null value on class JavaTestSubject."`